modules/local/pairtools/main.nf

```diff
@@ -0,0 +1,140 @@
+process PAIRTOOLS {
+    tag "${meta.id}"
+    label 'process_high'
+
+    // Pinning numpy to 1.23 until https://github.com/open2c/pairtools/issues/170 is resolved
+    // Not an issue with the biocontainers because they were built prior to numpy 1.24
+    conda "bioconda::pairtools=1.1.0 conda-forge::numpy=1.23"
+    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
+        'https://depot.galaxyproject.org/singularity/pairtools:1.1.0--py39hd5a99d8_2' :
+        'biocontainers/pairtools:1.1.0--py39hd5a99d8_2' }"
+
+    input:
+    tuple val(meta), path(bam)
+    path  chromsizes
+    path  fasta
+    val   sort_bam
+
+    output:
+    tuple val(meta), path("*.split.pairs.gz")   , emit: pairs
+    tuple val(meta), path("*.split.pairs.bam")  , emit: bam , optional:true
+    tuple val(meta), path("*.split.pairs.cram") , emit: cram, optional:true
+    tuple val(meta), path("*.crai")             , emit: crai, optional:true
+    tuple val(meta), path("*.csi")              , emit: csi , optional:true
+    path "*.stat"                               , emit: stat
+    path "versions.yml"                         , emit: versions
+
+    when:
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args = task.ext.args ?: ''
+    def args2 = task.ext.args2 ?: ''
+    def args3 = task.ext.args3 ?: ''
+    def args4 = task.ext.args4 ?: ''
+    def args5 = task.ext.args5 ?: ''
+    def args6 = task.ext.args6 ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def bamCollection = bam instanceof Collection ? bam : [bam]
+    def bamCount = bamCollection.size()
+    def maxParallel = Math.min(bamCount, task.cpus)
+    def buffer = task.memory.toGiga().intdiv(2)
+    def bufferPerJob = (buffer / maxParallel).max(1).toInteger()
+    def cpusPerJob = (task.cpus / maxParallel).max(1)
+    def samtools_command = sort_bam ? 'sort' : 'view'
+    def extension_pattern = /(--output-fmt|-O)+\s+(\S+)/
+    def extension_matcher =  (args6 =~ extension_pattern)
+    def extension = extension_matcher.getCount() > 0 ? extension_matcher[0][2].toLowerCase() : "bam"
+    def reference = fasta && extension=="cram"  ? "--reference ${fasta}" : ""
+    if (!fasta && extension=="cram") error "Fasta reference is required for CRAM output"
+    """
+    export MPLCONFIGDIR=tmp
+
+    # Multiple BAMs
+
+    if (( ${bamCount} > 1 )); then
```

Comment:
I think we're making this unnecessarily complicated like this (hard to read, etc). I think the `xargs` was a poor suggestion on my part. There's a lot of micromanaging of resources that we should just leave to the OS. 

I think we should stick to a for loop and allocate the full cpus and memory to the steps. 

```
BAM_FILES=($bam.join(' '))
TEMP_FILES=()

for BAM in "\${BAM_FILES[@]}"; do
    BAM_PREFIX=\$(basename "\$BAM" .bam)
    TEMP_FILE="\$BAM_PREFIX.pairs_temp.gz"
    TEMP_FILES+=("\$TEMP_FILE")
    pairtools parse ...
    pairtools sort ...
done

pairtools merge \\
    ... \\
    "\${TEMP_FILES[@]}" | \\
...

rm -f "\${TEMP_FILES[@]}"
```